aip/general/0236.md

```diff
@@ -0,0 +1,377 @@
+enforced.
+
+### Supported use-cases
+
+The following use-cases are considered when making a recommendation for API best
+practices of preview mode:
+- Resource: Each entity to be previewed is a gRPC resource.
+- Collection: A collection of objects live on a REST resource and one or many of
+these can be previewed.
+
+Each API recommended below will have a table explaining how it translates to
+each use-case.
+
+### Nested collection called experiments
+
+A new configuration of a Policy which is to be validated using preview mode is
+stored in a nested collection under that Policy resource.
+
+In this AIP, we will use the example of a hypothetical Policy resource called,
+`Policy`.
+
+`/projects/{project}/locations/{location}/policies/{policy}`
+
+The experimental versions of the resource used for preview mode are stored as a
+nested collection under `Policy`. The resource type will follow the convention
+"<resource_name> + Experiment”.
+
+`/projects/{project}/locations/{location}/policies/{policy}/experiments/{experiment}`
+
+A proto used to represent an experiment **must** contain the following fields:
+
+  1. The required top-level fields for a resource, like `name` and `etag`.
+  2. The `Policy` that is being tested itself.
+  3. The `metadata` field for preview.
+
+
+```proto
+message PolicyExperiment {
+
+  // The name of the PolicyExperiment.
+  string name = 1;
+ 
+  // The eTag of the PolicyExperiment.
+  string etag = 2;
+
+  // The fields used for log evaluation and eventual promotion.
+  Policy policy = 3;
+
+  // Metadata for preview mode only.
+  PreviewMetadata preview_metadata = 4 [(google.api.field_behavior) = OUTPUT_ONLY];
+}
+```
+- A new proto **must** be used for experiments with the Policy proto inlined
+  inside.
+  - The inlined Policy field contains any changes to the parent resource which
+    are to be tested using preview mode.
+- The name inside the embedded Policy message **must** be the name of the parent. 
+- When the user is ready to promote an experiment, the Policy message field
+  **must** be copied into the live resource and the experiment is deleted.
+  This can be done manually or via a custom method for commit described below.
+- A product **may** support multiple new configurations concurrently in preview
+  for a singular parent Policy. This is represented using two or more resources
+  in the nested collection.
+  - The number of configurations in preview for a given parent **may** be capped
+    at a certain number.
+- Cascading deletes **must** occur: if the top-level resource is deleted, all
+  resources underlying any experiments must also be deleted.
+- The etag on the experiment **must** guidance on the resource freshness
+  validation ([AIP-154][aip-154]) and will be a randomly generated UUID
+  ([AIP-148][aip-148]). The etag **must** be present in the preview logs.
+  - The parent resource **must** have an `etag` as well.
```

Comment:
Remove "an" or backticks. If "an" is there, you're talking about it as a term, if backticks are there, you're referring to it as symbol, i.e. the field name.